### DISCUSSION TITLE
Scoped const variables strange behavior

### DISCUSSION THREAD

**Discussion by mat100payette** [airsdk/Adobe-Runtime-Support#1541] [Q&A] [answered]
I have noticed that many of the `const` variables I declare (for the sake of telling other devs that it shouldnt be overwritten in scope) compile to unnamed locals.

For example,
```as
const myConst:int = 4 + parseInt("2");
```
would compile as
```as
var _loc1_:int;
_loc1_ = 4 + parseInt("2");
```
Looking at the disassembled code instructions, these assignments are done using the `setlocalN` opcode (in this case, `setlocal1` because it's `_loc1_`).

What really puzzles me though is that at runtime (on vscode) I cannot inspect these `const` variables since they're not given any debug label for some reason (i.e. no `debug "myConst", 1, 1` instruction is generated for it, whereas all `var` variables have one).

Am I stuck with never being able to properly debug scoped `const` variables ?

**Comment by mat100payette** [comment]
Please refer to this as3 debugger extension issue for more context:
https://github.com/BowlerHatLLC/vscode-swf-debug/issues/37

**Comment by mat100payette** [comment, marked as answer]
@ajwfrost After a lot of research and debugging, I have two hypothesis:

___
**_Hypothesis 1:_**

When the debugger asks for a frame data, I made sure that the output message is in the following format:
- DWORD 4 (msg size)
- DWORD 26 (type InFrame)
- 4 empty bytes

I'm assuming this is all correct, and that there is no missing data from this message to get the frame from the player.
If this is all correct, then the conclusion is that the player does not include all the local registers' data in the response message, only those that have a debug tag generated by the compiler.

In this scenario, I'd really like it if you could update that message so it also includes the other unnamed local registers by simply giving them a name that represents their index. That should be fully collision free since debug labels use compile-time variable names, which cannot start with a digit.

___
**_Hypothesis 2:_**

The other plausible option (although I believe it's more unlikely) is that the debugger maybe doesn't fully or properly parse the response messages from the player, resulting in unnamed local registers' data not being populated at all on our side. If you believe this is the case, could you please try to explain to us what message should be sent, or what response should be parsed, to get such missing data ?

**Reply by ajwfrost** [reply]
Hi @mat100payette  - this is a fascinating topic... not something we'd really had to look into before.

So, here's how it works:
- when the debugger is attached, and we go into a new function, the runtime sets up the local frame information based on the method info from the ActionScript bytecode, which tells us how many function args and other local variables there are
- slot zero is always "this"
- the next slots are taken up with parameter names (when embedded in the ABC, which they are if you build with -debug)
- the next slots are the rest of the local variables
- when the runtime then encounters a `debug` command, it uses this to track the name information for a local variable, for example if I see `debug 1 17 2 0` it means it's a `DI_LOCAL` instruction (to set a local variable name) with the name at index 17 of the string constants and at local variable number 2
- when the debugger requests the local frame information, the returned local variables are only included in this if they have had the name set up in the above fashion

So basically, you're completely correct with hypothesis 1:
> the player does not include all the local registers' data in the response message, only those that have a debug tag generated by the compiler.

I would guess the difference in behaviour between the Flex compiler and the AIR compiler is behind the change here, there are a number of differences although this one doesn't seem to bring a huge benefit if you're building in debug mode anyway. So a "real" fix would be to the compiler: we can add that to the (increasing) number of items we want to address in that ...

But as a short term fix, as you suggest, we can just change the "ignore this local variable if the name is undefined" to "include this local variable using a generated name". It may not help completely as you're not able to e.g. hover over the const variable to get its value, but it would at least show up in a list of local variables... see below...

![image](https://user-images.githubusercontent.com/6400804/148779798-171b7500-345e-4e3a-9d25-af414c642dfe.png)


Thanks for pointing this out!

**Reply by mat100payette** [reply]
Thanks the the in-depth reply!

I dont know if you have had time to fully read the issue I linked, but essentially I have a hacky solution to display those values on hover if the runtime at least gives the values with their local register index. Basically, I decompile the code at runtime (when stopped on a breakpoint or paused, or at program start), then I associate each `setlocal` to a source code line using the closest previous `debugline` instruction, and I parse that line from the source code to extract the field name. This way, when I hover above `myConst`, I can grab its correct local register value from the mapping I made.

So yeah, if it's quicker to make the simple runtime fix, that'd still help me a lot. It wouldnt be the most performant thing, but the mapping process isnt terribly slow either so it would be a good temporary solution until the compiler can be fixed.

**Reply by ajwfrost** [reply]
Ah okay, that's cunning! So if you can do that and then grab the value of the variable using the proposed formatting (`"%d (unnamed local variable)"`) then it should work...  we'll push this change into the next release (which should be out fairly soon...)

thanks

**Reply by mat100payette** [reply]
@ajwfrost Before I open an issue, may I ask you to please explain to me how you determine the unnamed variable index you're assigning ? They're somehow not consistent with the decompiling solution I had planned.

**Reply by ajwfrost** [reply]
Hi @mat100payette, this is just using the index of the variable in the method's "local" information. The debugger basically goes through all the local variables and calls `getLocalName()` and had been ignoring any names that were 'kundefined'. See https://github.com/adobe/avmplus/blob/master/core/MethodInfo.cpp - the values are set up via `setRegName` a little way below the `getLocalName` definition. We now basically do `String varname = getLocalName(i); if (varname == kundefined) // create new string from 'i'`

Note that there's an offset because of function arguments in case you're looking at this? So the "debug" commands have an index which would be into the registers, the first registers are used for the function parameters and the rest used for local variable names.

So if a function is defined as `myFunc(param1 : String) : void` and has a local variable `var myVal : int` then you would expect two `debug` commands:
```
  debug    1 22 0 0  // index 0 which is referring to the parameter
  debug    1 25 1 0  // index 1 which is referring to the local variable
```
where the values are then kept in the constant pool's string list at indices 22 and 25 for "myFunc" and "myVal"...

Hope that helps .. ?

thanks

**Reply by mat100payette** [reply]
@ajwfrost Hey thanks for the quick reply once again.

Yeah this cannot work. If it's not too much to ask, would you mind giving me the SWF file you test on (preferably one with a function that has both `const` and `var` definitions in it), and the assembly code you get from it ? I will have to compare that to what I get before I get into the details of how we both name the "unnamed" variables.

**Reply by mat100payette** [reply]
Actually, to save us some back and forth messages, here's what I observe with my SWF:

1. This is the source code of a function I debug:
![UnnamedSource](https://user-images.githubusercontent.com/20545324/150385240-ef291bfc-274d-42ca-9118-b2afcde32c12.png)

The important lines in this function are 135 and 136.

2. This is what the debugger shows with the latest SDK, when I'm stopped on a breakpoint at line 144:
![UnnamedLocals](https://user-images.githubusercontent.com/20545324/150385695-de1cd754-e321-4a05-88d8-41685d119e44.png)

By the values, I know that `1 (unnamed local variable)` is `preStringified`. So, I would expect to find that `preStringified` is put into local register 1, right ? Correct me if that's a wrong assumption.

3. Now, look at what assembly code I get from decompiling the file:
![UnnamedAssembly](https://user-images.githubusercontent.com/20545324/150388340-f88eb128-27b2-4812-aa04-c5c200845919.png)

Two things confuse me here. First thing is that `debug 1, "json", 2, 0` line. What I understand is that this should mean that the string lookup for `"json"` should grab the value from local register 2, right ?

However, if I look at the operations, it seems pretty clear that debugline 135 sets a value in local register 2, and that value is... the `preStringified` value. Moreover, debugline 136 (which sets `json`) assigns a value in local register 3 (???) but we already established that the `debug` statement said the lookup of `"json"` was in reg 2. So, am I misunderstanding something here ? Maybe the [decompiler](https://github.com/jindrapetrik/jpexs-decompiler/wiki/Library) I use is wrong ? I'm really unsure, but what I know is that we don't assign register index the same way.

**Reply by jindrapetrik** [reply]
The `debug` instructions third parameter is index of the register minus one. So 0 means register1, 1 means register2, etc.
Because register0 is always `this` variable.
In your case there is `debug 1, "json", 2, 0` which means it's register3 - setlocal3/getlocal3 is used in that case.

**Reply by mat100payette** [reply]
> The `debug` instructions third parameter is index of the register minus one. So 0 means register1, 1 means register2, etc. Because register0 is always `this` variable. In your case there is `debug 1, "json", 2, 0` which means it's register3 - setlocal3/getlocal3 is used in that case.

Omg wow that helps a lot to clear some of my confusion. I kinda wish the AVM2 manual explained that... You can see on page 52 of [the manual](https://www.adobe.com/content/dam/acom/en/devnet/pdf/avm2overview.pdf) that it says:

> _reg_ is an unsigned byte and is the index of the register that this is debugging information for.

which is clearly inaccurate given your explanation.

___

That being said, I believe my assumptions are correct then, and the latest update to the runtime to expose the unnamed variables is not working as expected (i.e. not exposing the proper reg index).

**Reply by ajwfrost** [reply]
Hi - sorry, missed the earlier updates a couple of days ago. I would agree here re. the offset being caused by the `this` variable which is put into the local `0` slot. If you look at the local variables as reported by the debugger, you can see they're in order and match up to what the `debug` instructions say (with the offset):

```
this                << special case, available via getlocal0
aa                  << debug slot 0, stored in setlocal1
1 = preStringified  << debug slot 1, stored in setlocal2
json                << debug slot 2, stored in setlocal3
stringified         << debug slot 3, stored in setlocal4
key                 << debug slot 4, stored in setlocal5
5 = unnamed temporary variable used during the function?
6 = unnamed temporary variable used during the function?
```

I don't know whether it's possible - but I would hope it is - that the `debug` commands could appear later on in a function and would then change use of a local slot. Maybe that would only happen in a release/optimised build, but I would hope if you did something like:
```
var i : int = something;
call(myArray[i]);
var j : int = somethingElse;
call(myArray[j]);
```
that there would be no need to use different local slots for `i` and `j`, they could both be stored in the same location as there's no need for `i` to be retained. Anyway, not sure whether that is relevant, but it might be worth bearing in mind.

But just to check: the whole thing you're doing here with the decompilation and working out variable names etc, is to support the debugging of constant values, and the reason why it doesn't work in the first place is just because the `debug` instructions aren't generated for constants? So all this would just go away if we could get the compiler to spit those out in the first place..?

thanks

**Reply by mat100payette** [reply]
@ajwfrost 

> So all this would just go away if we could get the compiler to spit those out in the first place..?

Yes absolutely ! I just didn't want to push too much on that aspect given your earlier comment:

> So a "real" fix would be to the compiler: we can add that to the (increasing) number of items we want to address in that ...

Having proper `debug` commands generated for `const` would be the ideal goal. All the convoluted solution I prepared was in case the modifications to the compiler would be difficult and/or take a long time to be implemented.

Thanks again for the great collaboration and discussion, I appreciate the help and effort.

**Reply by mat100payette** [reply]
Just so you know, with that offset info I managed to update my vscode debugger extension fork and here's the result:
![WorkingConst](https://user-images.githubusercontent.com/20545324/150666568-bb999e91-3140-4950-ae82-77cec6e052a6.png)

Local `const` displays in variables panel and can be hovered for value.

As for the mystery `4` and `5` unnamed variables, I can tell from their values that they are a class variable (`_compatSettings`) and a class method closure (`replacer`) respectively (disclaimer I'm not 100% certain about the latter). These are simply not yet parsed away by the debugger extension; that shouldn't be too hard though because they match members of the `this` variable. So basically this hacky solution is done, but if you can give me some sort of ETA on the inclusion of `debug` instructions directly from the compiler for `const`, that'd give me an idea of whether or not I should clean this hack and PR it.

**Reply by ajwfrost** [reply]
@mat100payette well looking at it .. and using the Royale compiler which is pretty similar to what we have from Adobe:
https://github.com/apache/royale-compiler/blob/develop/compiler/src/main/java/org/apache/royale/compiler/internal/as/codegen/LexicalScope.java
Lines 1435-1436 appear to be the culprits i.e. the `continue` in the below:
```
            for (Binding local_binding : localBindings.values())
            {
                if (SemanticUtils.isConstDefinition(local_binding.getDefinition()))
                    continue;

                // Need to -1 the index, as the local register starts at 1 to handle the this
                // but that's not included in the bindings list
                int index = local_binding.getLocalRegister() - 1;
                addDebugNameToDefinition(local_binding.getName().getBaseName(), index, result);
            }
```
I'm slightly concerned at the possibility of side-effects but we can have a bit of a play to see what changes. But really as long as the local register is a valid number, I can't see a big problem...

thanks

**Reply by ajwfrost** [reply]
@mat100payette can you try with the below please? I'm a little surprised that it's this easy but I think this may just help....
thanks
https://transfer.harman.com/message/DYa9GLLvdMijvpcqg9CP6k

**Reply by mat100payette** [reply]
@ajwfrost Will try in a bit. I wish you guys were open source haha, I could most likely have found that out and saved both of us all this work loll.

**Reply by mat100payette** [reply]
@ajwfrost I can confirm that `debug` instructions can be seen for all `const` variables I've checked so far in my codebase. That's awesome. I would imagine that it means you can rollback the changes to expose "unnamed variables" now then. Small question: is there any reason why `debug` instructions aren't ordered by register or name ? I feel like register ordering (and then maybe name ordering on top of it) would be the cleanest. Just a detail though.

Thanks for the work, we seem to have reached its conclusion haha.

**Reply by ajwfrost** [reply]
Hi @mat100payette  - thanks for checking that. Not sure why you're saying the `debug` instructions aren't ordered: everywhere I've seen, they've been in slot order:
```
       4    debug         	1 109 0 0
       9    debug         	1 110 1 0
       14   debug         	1 111 2 0
       19   debug         	1 112 3 0
```
although I guess looking at the code, it just depends on what order that loop happens in i.e. `for (Binding local_binding : localBindings.values())`. No particular reason why these couldn't then be ordered within that method but I'm not sure why it would be necessary...?

thanks

**Reply by mat100payette** [reply]
@ajwfrost Really just QoL for when I look at disassembled code. It's no big deal at all haha.
